packages/xstate-react/src/useActor.ts

```diff
@@ -1,69 +1,45 @@
-import { useState, useEffect, useMemo, useRef, useCallback } from 'react';
-import { ActorRef, Sender, ActorRefLike } from './types';
-import { EventObject, Interpreter, Actor } from 'xstate';
-import { fromService } from './useService';
-
-function resolveActor(
-  actorLike:
-    | ActorRef<any, any>
-    | ActorRefLike<any, any>
-    | Interpreter<any, any>
-    | Actor
-): ActorRef<any, any> & { deferred?: boolean } {
-  if (actorLike instanceof Interpreter) {
-    return fromService(actorLike);
-  }
-
-  if (!('current' in actorLike)) {
-    return {
-      stop: () => {
-        /* do nothing */
-      },
-      ...actorLike,
-      subscribe: actorLike.subscribe as any, // TODO: fix (subscribe will work, just improperly typed)
-      name: actorLike.id!,
-      current: actorLike.state!
-    };
-  }
-
-  return actorLike;
-}
+import { useState, useEffect, useRef, useCallback } from 'react';
+import { Sender, ActorRefLike } from './types';
+import { EventObject, Actor } from 'xstate';
 
 export function useActor<TEvent extends EventObject, TEmitted = any>(
-  actorLike: ActorRefLike<TEvent, TEmitted> | Actor
+  actorRef: ActorRefLike<TEvent, TEmitted> | Actor,
+  getSnapshot: (actor: typeof actorRef) => TEmitted = (a) =>
+    'state' in a ? a.state : (undefined as any)
 ): [TEmitted, Sender<TEvent>] {
-  const actor = useMemo(() => resolveActor(actorLike), [actorLike]);
-  const deferredEventsRef = useRef<EventObject[]>([]);
-  const [current, setCurrent] = useState(actor.current);
+  // const actor = useMemo(() => resolveActor(actorLike), [actorLike]);
+  const deferredEventsRef = useRef<TEvent[]>([]);
+  const [current, setCurrent] = useState(() => getSnapshot(actorRef));
 
   const send: Sender<TEvent> = useCallback(
     (event) => {
       // If the previous actor is a deferred actor,
       // queue the events so that they can be replayed
       // on the non-deferred actor.
-      if (actor.deferred) {
+      if ('deferred' in actorRef && actorRef.deferred) {
         deferredEventsRef.current.push(event);
       } else {
-        actor.send(event);
+        actorRef.send(event);
       }
     },
-    [actor]
+    [actorRef]
   );
 
   useEffect(() => {
-    const subscription = actor.subscribe(setCurrent);
+    setCurrent(getSnapshot(actorRef));
+    const subscription = actorRef.subscribe(setCurrent);
```

Comment:
shouldnt this be
```suggestion
    const subscription = actorRef.subscribe(() => setCurrent(getSnapshot(actorRef)));
```
?

Reply:
No it shouldn't; `getSnapshot` provides the last known snapshot of the last emitted value, while `subscribe` is the _actual_ latest emitted value.

In other words, `.subscribe(value => ...)` should be the ultimate source of truth. `getSnapshot(...)` is useful for the initial state.

Reply:
oh, ok - thought it was about selecting a slice of the emitted values, hence the confusion

👍

Reply:
Selecting a slice is for the next PR 😉